README.md
# Luacheck

[![Join the chat at https://gitter.im/luacheck/Lobby](https://badges.gitter.im/luacheck/Lobby.svg)](https://gitter.im/luacheck/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![Luacheck](https://img.shields.io/github/workflow/status/lunarmodules/luacheck/Luacheck?label=Luacheck&logo=Lua)](https://github.com/lunarmodules/luacheck/actions?workflow=Luacheck)
[![Busted](https://img.shields.io/github/workflow/status/lunarmodules/luacheck/Busted?label=Busted&logo=Lua)](https://github.com/lunarmodules/luacheck/actions?workflow=Busted)
[![Coverage Status](https://img.shields.io/coveralls/github/lunarmodules/luacheck?label=Coveralls&logo=Coveralls)](https://coveralls.io/github/lunarmodules/luacheck?branch=master)
[![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/lunarmodules/luacheck?label=Tag&logo=GitHub)](https://github.com/lunarmodules/luacheck/releases)
[![Luarocks](https://img.shields.io/luarocks/v/lunarmodules/luacheck?label=Luarocks&logo=Lua)](https://luarocks.org/modules/lunarmodules/luacheck)

## Contents

* [Overview](#overview)
* [Installation](#installation)
* [Basic usage](#basic-usage)
* [Related projects](#related-projects)
* [Documentation](#documentation)
* [Development](#development)
* [Building and testing](#building-and-testing)
* [License](#license)

## Overview

Luacheck is a static analyzer and a linter for [Lua](http://www.lua.org). Luacheck detects various issues such as usage of undefined global variables, unused variables and values, accessing uninitialized variables, unreachable code and more. Most aspects of checking are configurable: there are options for defining custom project-related globals, for selecting set of standard globals (version of Lua standard library), for filtering warnings by type and name of related variable, etc. The options can be used on the command line, put into a config or directly into checked files as Lua comments.

Luacheck supports checking Lua files using syntax of Lua 5.1 - 5.4, and LuaJIT. Luacheck itself is written in Lua and runs on all of mentioned Lua versions.

## Installation

### Using LuaRocks

From your command line run the following command (using `sudo` if necessary):

```
luarocks install luacheck
```

For parallel checking Luacheck additionally requires [LuaLanes](https://github.com/LuaLanes/lanes), which can be installed using LuaRocks as well (`luarocks install lanes`).

### Windows binary download

For Windows there is single-file 64-bit binary distribution, bundling Lua 5.4.4, Luacheck, LuaFileSystem, and LuaLanes using [LuaStatic](https://github.com/ers35/luastatic):
[download](https://github.com/lunarmodules/luacheck/releases/download/v1.0.0/luacheck.exe).

## Basic usage

After Luacheck is installed, run `luacheck` program from the command line. Pass a list of files, [rockspecs](https://github.com/luarocks/luarocks/wiki/Rockspec-format) or directories (requires LuaFileSystem) to be checked:

```
luacheck src extra_file.lua another_file.lua
```

```
Checking src/good_code.lua               OK
Checking src/bad_code.lua                3 warnings

    src/bad_code.lua:3:23: unused variable length argument
    src/bad_code.lua:7:10: setting non-standard global variable embrace
    src/bad_code.lua:8:10: variable opt was previously defined as an argument on line 7

Checking src/python_code.lua             1 error

    src/python_code.lua:1:6: expected '=' near '__future__'

Checking extra_file.lua                  5 warnings

    extra_file.lua:3:18: unused argument baz
    extra_file.lua:4:8: unused loop variable i
    extra_file.lua:13:7: accessing uninitialized variable a
    extra_file.lua:14:1: value assigned to variable x is unused
    extra_file.lua:21:7: variable z is never accessed

Checking another_file.lua                2 warnings

    another_file.lua:2:7: unused variable height
    another_file.lua:3:7: accessing undefined variable heigth

Total: 10 warnings / 1 error in 5 files
```

For more info, see [documentation](https://luacheck.readthedocs.io/en/stable/).

## Related projects

### Editor support

There are a few plugins which allow using Luacheck directly inside an editor, showing warnings inline:

* For Vim, [Syntastic](https://github.com/vim-syntastic/syntastic) contains [luacheck checker](https://github.com/vim-syntastic/syntastic/wiki/Lua%3A---luacheck);
* For Sublime Text 3 there is [SublimeLinter-luacheck](https://packagecontrol.io/packages/SublimeLinter-luacheck) which requires [SublimeLinter](https://sublimelinter.readthedocs.io/en/latest/);
* For Atom there is [linter-luacheck](https://atom.io/packages/linter-luacheck) which requires [AtomLinter](https://github.com/steelbrain/linter);
* For Emacs, [Flycheck](http://www.flycheck.org/en/latest/) contains [luacheck checker](http://www.flycheck.org/en/latest/languages.html#lua);
* For Brackets, there is [linter.luacheck](https://github.com/Malcolm3141/brackets-luacheck) extension;
* For Visual Studio code there is [vscode-luacheck](https://marketplace.visualstudio.com/items?itemName=dwenegar.vscode-luacheck) extension. [vscode-lua](https://marketplace.visualstudio.com/items?itemName=trixnz.vscode-lua) extension also includes Luacheck support;
* For Nova, search the Extension Library for the [Luacheck](https://github.com/GarrettAlbright/Luacheck.novaextension) extension.

If you are a plugin developer, see [recommended way of using Luacheck in a plugin](http://luacheck.readthedocs.org/en/stable/cli.html#stable-interface-for-editor-plugins-and-tools).

### Other projects

* [Luacheck bindings for Node.js](https://www.npmjs.com/package/luacheck);
* [Luacheck plugin for Gulp](https://www.npmjs.com/package/gulp-luacheck).

## Documentation

Documentation is available [online](https://luacheck.readthedocs.io/en/stable/). If Luacheck has been installed using LuaRocks, it can be browsed offline using `luarocks doc luacheck` command.

Documentation can be built using [Sphinx](http://sphinx-doc.org/): `sphinx-build docsrc doc`, the files will be found inside `doc/`.

## Development

Luacheck is currently in development. The latest released version is v1.0.0. The interface of the `luacheck` module may change between minor releases. The command line interface is fairly stable.

Use the Luacheck issue tracker on GitHub to submit bugs, suggestions and questions. Any pull requests are welcome, too.

## Building and testing

After the Luacheck repo is cloned and changes are made, run `luarocks make` (using `sudo` if necessary) from its root directory to install dev version of Luacheck. To run Luacheck using sources in current directory without installing it, run `lua -e 'package.path="./src/?.lua;./src/?/init.lua;"..package.path' bin/luacheck.lua ...`. To test Luacheck, ensure that you have [busted](http://olivinelabs.com/busted/) and [luautf8](https://github.com/starwing/luautf8) installed and run `busted`.

## Docker

Alternatively Luacheck can be run as a standalone docker container.
The usage of docker is fairly simple.
You can either build your own or download a prebuilt version.
To build your own, execute the following command from the source directory of this project:

```console
$ docker build -t ghcr.io/lunarmodules/luacheck:HEAD .
```

To use a prebuilt one, download it from the GitHub Container Registry.
Here we use the one tagged *latest*, but you can substitute *latest* for any tagged release.

```console
$ docker pull ghcr.io/lunarmodules/luacheck:latest
```

Once you have a container you can run it on one file or a source tree (substitute *latest* with *HEAD* if you built your own or with the tagged version you want if applicable):

```console
# Run on an entire directory
$ docker run -v "$(pwd):/data" ghcr.io/lunarmodules/luacheck:latest .

# Run on one file:
$ docker run -v "$(pwd):/data" ghcr.io/lunarmodules/luacheck:latest bin/luacheck.lua
```

A less verbose way to run it in most shells is with at alias:

```console
# In a shell or in your shell's RC file:
$ alias luacheck='docker run -v "$(pwd):/data" ghcr.io/lunarmodules/luacheck:latest'

# Thereafter just run:
$ luacheck .
```
### Use as a CI job

There are actually many ways to run Luacheck remotely as part of a CI work flow.
Because packages are available for many platforms, one way would be to just use your platforms native package installation system to pull them into whatever CI runner environment you already use.
Another way is to pull in the prebuilt Docker container and run that.

As a case study, here is how a workflow could be setup in GitHub Actions:

```yaml
name: Luacheck
on: [push, pull_request]
jobs:
  sile:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Luacheck linter
        uses: lunarmodules/luacheck@v0
```

By default the GH Action is configured to run `luacheck .`, but you can also pass it your own `args` to replace the default input of `.`.

```yaml
      - name: Luacheck linter
        uses: lunarmodules/luacheck@v0
        with:
            args: myfile.lua
```

## License

```
The MIT License (MIT)

Copyright (c) 2014 - 2018 Peter Melnichenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v1.0.0 (2022-08-24)

### Documentation

- [**breaking**] Follow semver guidelines, next release will be v1.x.y

### Features

- Overhaul docker container to run on Lua 5.4
- Store cached luacheck values per-version in case of changes
- Set_default_std for ldoc
- Add builtin std option for Ldoc globals
- Add builtin std option for the Playdate SDK

## v0.26.1 (2022-04-23)

### Bug Fixes

- Exempt special builtin \_ENV from 214 warning

### Features

- In case of no home environment, default to caching in CWD (#60)
- Add multi-thread support to container (#59)

### Miscellaneous Tasks

- Tweak warning message for 214 to be more explicit

## v0.26.0 (2022-03-23)

### New Features

* Add "unused hint" flag — @javierguerragiraldez
  Function arguments that start with a single underscore get an "unused hint".
  Leaving them unused doesn't result in a warning.
  Using them, on the other hand, is a new warning (№ 214).
* Add hook to enable Luacheck for use in pre-commit — @mblayman
* Warn on error-prone and unnecessary negations — @arichard4
  Two new warnings (№ 581 and 582) flag error-prone operator orders.
* Add Dockerfile implementation — @MartinBroers
  A 6.34MB containerized image with everything needed to run the linter.
  Build your own or pull prebuilt images from GHRC.
* Setup repository for use *as* a GitHub Action — @alerque
  Lint your repositories using GitHub Action workflows with just a single `uses:` step.

### Fixes

* Don't mark variables as accessed if only references are circular — @arichard4
* Make test suite Lua 5.4 compatible — @alerque
* Correct small issues in documentation — various

### Miscellaneous

* Overhaul CI workflows for testing, linting, building, and releasing — @alerque
* Update URLs and documentation reflecting new repository home — @lunarmodules

## 0.25.0 (2020-08-25)

### New features

* New values for CLI option --std: lua54, lua54c
* New lua54 standard library definitions

### Fixes

* Lua 5.4 allows 31bits utf8 codepoint

## 0.24.0 (2020-08-20)

### Changes

* Caching now uses files in a global directory instead of local
  `.luacheckcache` file. Default cache directory is
  `%LOCALAPPDATA%\Luacheck\Cache` on Windows,
  `~/Library/Caches/Luacheck` on OS X/macOS, and
  `$XDG_CACHE_HOME/luacheck` or `~/.config/luacheck` on other systems.

### Fixes

* Fixes for using Luacheck with Lua 5.4:
  - Parse 5.4 attributes in `local` declarations (but ignores them for now)
  - Luacheck itself can also run with Lua 5.4
* Fixed `randomize` missing from `busted` set of standard globals (#183).
* Added additional `table` and `thread` definitions for `ngx_lua`.
* Added additional `match` definition for `busted`.

### Miscellaneous

* Upgraded Windows binary components: Lua 5.3.4 -> 5.3.5,
  LuaFileSystem 1.6.3 -> 1.7.0.

## 0.23.0 (2018-09-18)

### Breaking changes

* Removed `--no-inline` CLI option and `inline` config option, inline options
  are now always enabled.
* Inline comments are now supposed to be only in short comments
  but not long ones.
* Installer script (install.lua) is removed. Luacheck can still be installed
  manually by recursively copying `src/*` to a directory in `package.path`
  and copying `bin/luacheck.lua` to a directory in `PATH` as `luacheck`.

### New features and improvements

* Warning columns are now reported in Unicode codepoints if input is
  valid UTF-8 (#45).
* Added indentaion-based guessing of a better location for missing `end`
  and `until` syntax errors.
* Added `luacheckrc` set of allowed globals containing globals used in
  Luacheck config to set options.
* Added default stds equivalent to predefined per-path std overrides
  in config:
  - `files["**/spec/**/*_spec.lua"].std = "+busted"`;
  - `files["**/test/**/*_spec.lua"].std = "+busted"`;
  - `files["**/tests/**/*_spec.lua"].std = "+busted"`;
  - `files["**/*.rockspec"].std = "+rockspec"`;
  - `files["**/*.luacheckrc"].std = "+luacheckrc"`.
* Added detection of numeric for loops going from `#t` to `1` without
  negative step (#160).
* Added support for LuaRocks 3 module autodetection when checking
  rockspecs (#176).
* Updated `love` standard for LÖVE 11.1 (#178).

### Changes

* Default set of standard globals is now always `max`, allowing globals of all
  Lua versions. `_G` std is deprecated.

### Fixes

* Added missing globals to `rockspec` std: `hooks`, `deploy`,
  `build_dependencies`, `test_dependencies`, and `test`.
* Fixed line lengths appearing in the output before other warnings on the same
  line even if their column numbers are smaller.

### Miscellaneous

* Luacheck now depends on argparse instead of bundling it.
* LuaFileSystem dependency is now required.

## 0.22.1 (2018-07-01)

### Improvements

* Reduced amount of RAM used when checking a large number of files.

## 0.22.0 (2018-05-09)

### New features and improvements

* Added detection of cyclomatic complexity, with warnings emitted for
  functions with complexity higher than a configurable limit; disabled
  by default (#141).
* Added a built-in formatter printing warnings and errors in a format
  understood by MSBuild/Visual Studio if `luacheck` is
  used as a custom build step (#142).
* `ranges` and `quiet` options can now be used in config, e.g. `quiet = 1` to
  disable `OK` lines in default formatter output.
* `luacheck` module now adds `prev_line`, `prev_column`, and `prev_end_column`
  fields to syntax error events if they refer to some extra location:
  redefined label errors point to the previous definition,
  unpaired tokens such as `function`/`end` point to the the first token (#134).
* `luacheck` module now adds `prev_end_column` field to warning events that
  already have `prev_line` and `prev_column` fields, and
  `overwritten_end_column` for warnings with `overwritten_line` and
  `overwritten_column`.
* Improved error messages for invalid options and config: when an option is

src/luacheck/main.lua
local argparse = require "argparse"
local cache = require "luacheck.cache"
local config = require "luacheck.config"
local luacheck = require "luacheck"
local multithreading = require "luacheck.multithreading"
local profiler = require "luacheck.profiler"
local runner = require "luacheck.runner"
local utils = require "luacheck.utils"
local version = require "luacheck.version"

local exit_codes = {
   ok = 0,
   warnings = 1,
   errors = 2,
   fatals = 3,
   critical = 4
}

local function critical(msg)
   io.stderr:write("Critical error: "..msg.."\n")
   os.exit(exit_codes.critical)
end

local function get_parser()
   local parser = argparse(
      "luacheck", "luacheck " .. luacheck._VERSION .. ", a linter and a static analyzer for Lua.", [[
Links:

   Luacheck on GitHub: https://github.com/lunarmodules/luacheck
   Luacheck documentation: https://luacheck.readthedocs.org]])
      :help_max_width(80)

   parser:argument("files", "List of files, directories and rockspecs to check. Pass '-' to check stdin.")
      :args "+"
      :argname "<file>"

   parser:group("Options for filtering warnings",
      parser:flag("-g --no-global", "Filter out warnings related to global variables. " ..
         "Equivalent to --ignore 1."):target("global"):action("store_false"),
      parser:flag("-u --no-unused", "Filter out warnings related to unused variables and values. " ..
         "Equivalent to --ignore [23]."):target("unused"):action("store_false"),
      parser:flag("-r --no-redefined", "Filter out warnings related to redefined variables. " ..
         "Equivalent to --ignore 4."):target("redefined"):action("store_false"),

      parser:flag("-a --no-unused-args", "Filter out warnings related to unused arguments and " ..
         "loop variables. Equivalent to --ignore 21[23]."):target("unused_args"):action("store_false"),
      parser:flag("-s --no-unused-secondaries", "Filter out warnings related to unused variables set " ..
         "together with used ones."):target("unused_secondaries"):action("store_false"),
      parser:flag("--no-self", "Filter out warnings related to implicit self argument.")
         :target("self"):action("store_false"),

      parser:option("--ignore -i", "Filter out warnings matching these patterns.\n" ..
         "If a pattern contains slash, part before slash matches warning code and part after it matches name of " ..
         "related variable. Otherwise, if the pattern contains letters or underscore, it matches name of " ..
         "related variable. Otherwise, the pattern matches warning code.")
         :args "+"
         :count "*"
         :argname "<patt>"
         :action "concat"
         :init(nil),
      parser:option("--enable -e", "Do not filter out warnings matching these patterns.")
         :args "+"
         :count "*"
         :argname "<patt>"
         :action "concat"
         :init(nil),
      parser:option("--only -o", "Filter out warnings not matching these patterns.")
         :args "+"
         :count "*"
         :argname "<patt>"
         :action "concat"
         :init(nil),
      parser:option("--operators", "Allow compound operators matching patterns")
         :args "+"
         :count "*"
         :argname "<patt>"
         :action "concat"
         :init(nil))

   parser:group("Options for configuring allowed globals",
      parser:option("--std", "Set standard globals, default is max. <std> can be one of:\n" ..
         "   max - union of globals of Lua 5.1, Lua 5.2, Lua 5.3 and LuaJIT 2.x;\n" ..
         "   min - intersection of globals of Lua 5.1, Lua 5.2, Lua 5.3 and LuaJIT 2.x;\n" ..
         "   lua51 - globals of Lua 5.1 without deprecated ones;\n" ..
         "   lua51c - globals of Lua 5.1;\n" ..
         "   lua52 - globals of Lua 5.2;\n" ..
         "   lua52c - globals of Lua 5.2 with LUA_COMPAT_ALL;\n" ..
         "   lua53 - globals of Lua 5.3;\n" ..
         "   lua53c - globals of Lua 5.3 with LUA_COMPAT_5_2;\n" ..
         "   lua54 - globals of Lua 5.4;\n" ..
         "   lua54c - globals of Lua 5.4 with LUA_COMPAT_5_3;\n" ..
         "   luajit - globals of LuaJIT 2.x;\n" ..
         "   ngx_lua - globals of Openresty lua-nginx-module 0.10.10, including standard LuaJIT 2.x globals;\n" ..
         "   love - globals added by LÖVE;\n" ..
         "   playdate - globals added by the Playdate SDK;\n" ..
         "   busted - globals added by Busted 2.0, by default added for files ending with _spec.lua within spec, " ..
         "test, and tests subdirectories;\n" ..
         "   rockspec - globals allowed in rockspecs, by default added for files ending with .rockspec;\n" ..
         "   luacheckrc - globals allowed in Luacheck configs, by default added for files ending with .luacheckrc;\n" ..
         "   none - no standard globals.\n\n" ..
         "Sets can be combined using '+'. Extra sets can be defined in config by " ..
         "adding to `stds` global in config."),
      parser:flag("-c --compat", "Equivalent to --std max."),

      parser:option("--globals", "Add custom global variables (e.g. foo) or fields (e.g. foo.bar) " ..
         "on top of standard ones.")
         :args "*"
         :count "*"
         :argname "<name>"
         :action "concat"
         :init(nil),
      parser:option("--read-globals", "Add read-only global variables or fields.")
         :args "*"
         :count "*"
         :argname "<name>"
         :action "concat"
         :init(nil),
      parser:option("--new-globals", "Set custom global variables or fields. Removes custom globals added previously.")
         :args "*"
         :count "*"
         :argname "<name>"
         :action "concat"
         :init(nil),
      parser:option("--new-read-globals", "Set read-only global variables or fields. " ..
         "Removes read-only globals added previously.")
         :args "*"
         :count "*"
         :argname "<name>"
         :action "concat"
         :init(nil),
      parser:option("--not-globals", "Remove custom and standard global variables or fields.")
         :args "*"
         :count "*"
         :argname "<name>"
         :action "concat"
         :init(nil),

      parser:flag("-d --allow-defined", "Allow defining globals implicitly by setting them."),
      parser:flag("-t --allow-defined-top",
         "Allow defining globals implicitly by setting them in the top level scope."),
      parser:flag("-m --module", "Limit visibility of implicitly defined globals to their files."))

   parser:group("Options for configuring line length limits",
      parser:option("--max-line-length", "Set maximum allowed line length (default: 120).")
         :argname "<length>"
         :convert(tonumber),
      parser:flag("--no-max-line-length", "Do not limit line length.")
         :action "store_false"
         :target "max_line_length",

      parser:option("--max-code-line-length", "Set maximum allowed length for lines ending with code (default: 120).")
         :argname "<length>"
         :convert(tonumber),
      parser:flag("--no-max-code-line-length", "Do not limit code line length.")
         :action "store_false"
         :target "max_code_line_length",

      parser:option("--max-string-line-length", "Set maximum allowed length for lines within a string (default: 120).")
         :argname "<length>"
         :convert(tonumber),
      parser:flag("--no-max-string-line-length", "Do not limit string line length.")
         :action "store_false"
         :target "max_string_line_length",

      parser:option("--max-comment-line-length", "Set maximum allowed length for comment lines (default: 120).")
         :argname "<length>"
         :convert(tonumber),
      parser:flag("--no-max-comment-line-length", "Do not limit comment line length.")
         :action "store_false"
         :target "max_comment_line_length")

   parser:option("--max-cyclomatic-complexity", "Set maximum cyclomatic complexity for functions.")
      :argname "<complexity>"
      :convert(tonumber)
   parser:flag("--no-max-cyclomatic-complexity", "Do not limit function cyclomatic complexity (default).")
      :action "store_false"
      :target "max_cyclomatic_complexity"

   local default_global_path = config.get_default_global_path()

   local config_opt = parser:option("--config", "Path to configuration file. (default: "..config.default_path..")")
   local no_config_opt = parser:flag("--no-config", "Do not look up configuration file.")
      :action "store_false"
      :target "config"

   parser:mutex(config_opt, no_config_opt)

   local default_config_opt = parser:option("--default-config", ("Path to configuration file to use if --[no-]config "..
      "is not used and project-specific %s is not found. (default: %s)"):format(
         config.default_path, default_global_path or "could not detect"))
   local no_default_config_opt = parser:flag("--no-default-config", "Do not use default configuration file.")
      :action "store_false"
      :target "default_config"

   parser:mutex(default_config_opt, no_default_config_opt)

   parser:group("Configuration file options",
      config_opt,
      no_config_opt,
      default_config_opt,
      no_default_config_opt)

   parser:group("File filtering options",
      parser:option("--exclude-files", "Do not check files matching these globbing patterns.")
         :args "+"
         :count "*"
         :argname "<glob>"
         :action "concat"
         :init(nil),
      parser:option("--include-files", "Do not check files not matching these globbing patterns.")
         :args "+"
         :count "*"
         :argname "<glob>"
         :action "concat"
         :init(nil))

   parser:option("--filename", "Use another filename in output and for selecting configuration overrides.")

   local cache_opt = parser:option("--cache", ("Path to cache directory. (default: %s)"):format(
         cache.get_default_dir()))
      :args "?"

   local no_cache_opt = parser:flag("--no-cache", "Do not use cache.")
      :action "store_false"
      :target "cache"

   parser:mutex(cache_opt, no_cache_opt)

   local lanes_notice = ""

   if not multithreading.has_lanes then
      lanes_notice = "\nWarning: LuaLanes not found, parallel checking disabled."
   end

   parser:group("Performance optimization options",
      cache_opt,
      no_cache_opt,
      parser:option(
         "-j --jobs", "Check <jobs> files in parallel (default: " ..
            tostring(multithreading.default_jobs) .. ")." .. lanes_notice):convert(tonumber))

   parser:group("Output formatting options",
      parser:option("--formatter" , "Use custom formatter. <formatter> must be a module name or one of:\n" ..
         "   TAP - Test Anything Protocol formatter;\n" ..
         "   JUnit - JUnit XML formatter;\n" ..
         "   visual_studio - MSBuild/Visual Studio aware formatter;\n" ..
         "   plain - simple warning-per-line formatter;\n" ..
         "   default - standard formatter."),
      parser:flag("-q --quiet", "Suppress output for files without warnings.\n" ..
         "-qq: Suppress output of warnings.\n" ..
         "-qqq: Only print total number of warnings and errors.")
         :count "0-3",
      parser:flag("--codes", "Show warning codes."),
      parser:flag("--ranges", "Show ranges of columns related to warnings."),
      parser:flag("--no-color", "Do not color output.")
         :action "store_false"
         :target "color")

   parser:flag("--profile", "Show performance statistics."):hidden(true)

   parser:flag("-v --version", "Show version info and exit.")
      :action(function() print(version.string) os.exit(exit_codes.ok) end)

   return parser
end

local function main()
   local parser = get_parser()
   local ok, args = parser:pparse()
   if not ok then
      io.stderr:write(("%s\n\nError: %s\n"):format(parser:get_usage(), args))
      os.exit(exit_codes.critical)
   end

   if args.profile then
      args.jobs = 1
      profiler.init()
   end

   if args.quiet == 0 then
      args.quiet = nil
   end

   if args.cache then
      args.cache = args.cache[1] or true
   end

   local checker, err, is_invalid_args_error = runner.new(args)

   if not checker then
      if is_invalid_args_error then
         io.stderr:write(("%s\n\nError: %s\n"):format(parser:get_usage(), err))
         os.exit(exit_codes.critical)
      else
         critical(err)
      end
   end

   local inputs = {}

   for _, file in ipairs(args.files) do
      local input = {filename = args.filename}

      if file == "-" then
         input.file = io.stdin
      elseif file:find("%.rockspec$") then
         input.rockspec_path = file
      else
         input.path = file
      end

      table.insert(inputs, input)
   end

   local report, check_err = checker:check(inputs)

   if not report then
      critical(check_err)
   end

   for _, file_report in ipairs(report) do
      if not file_report.filename then
         file_report.filename = "stdin"
      end
   end

   local output, format_err = checker:format(report)

   if not output then
      critical(format_err)
   end

   io.stdout:write(output)

   if args.profile then
      profiler.report()
   end

   if report.fatals > 0 then
      os.exit(exit_codes.fatals)
   elseif report.errors > 0 then
      os.exit(exit_codes.errors)
   elseif report.warnings > 0 then
      os.exit(exit_codes.warnings)
   else
      os.exit(exit_codes.ok)
   end
end

local _, error_wrapper = utils.try(main)
local err = error_wrapper.err
local traceback = error_wrapper.traceback

if utils.is_instance(err, utils.InvalidPatternError) then
   critical(("Invalid pattern '%s'"):format(err.pattern))
elseif type(err) == "string" and err:find("interrupted!$") then
   critical("Interrupted")
else
   local msg = ("Luacheck %s bug (please report at https://github.com/lunarmodules/luacheck/issues):\n%s\n%s"):format(
      luacheck._VERSION, err, traceback)
   critical(msg)
end
